### PR TITLE
Add CORSAllowedOrigins Field to KubeAPIServerConfig

### PR DESCRIPTION
**Pull request by databus23**
**How to categorize this PR?**
/area control-plane
/kind api-change

**What this PR does / why we need it**:
I would like to be able to consume the Shoot kubernetes API directly from a browser.
This requires setting the flag `--cors-allowed-origins` in the kubernetes apiserver which is not exposed in the Shoot Spec so far.

**Special notes for your reviewer**:
I'm not familiar with the general gardener development workflow. If I missed something please let me know what else is needed to add a field to to Shoot API.

**Release note**:

```feature user
Allow to configure allowed CORS origins for shoot apiserver
```

### CODE REVIEW

File: pkg/apis/core/v1alpha1/types_shoot.go [hunk 1]
```diff
@@ -711,6 +711,10 @@ type KubeAPIServerConfig struct {
 	// Defaults to 300.
 	// +optional
 	DefaultUnreachableTolerationSeconds *int64 `json:"defaultUnreachableTolerationSeconds,omitempty" protobuf:"varint,15,opt,name=defaultUnreachableTolerationSeconds"`
+
+	//CORSAllowedOrigins contains a list of allowed origins for CORS, comma separated. An allowed origin can be a regular expression to support subdomain matching. If this list is empty CORS will not be enabled. (flag `--cors-allowed-origins`)
```
Comment (rfranzke) [inline pkg/apis/core/v1alpha1/types_shoot.go]:
```suggestion
	// CORSAllowedOrigins contains a list of allowed origins for CORS, comma separated. An allowed origin can be a regular expression to support subdomain matching. If this list is empty CORS will not be enabled. (flag `--cors-allowed-origins`)
```

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 2]
```diff
@@ -723,6 +723,10 @@ type KubeAPIServerConfig struct {
 	// Defaults to 300.
 	// +optional
 	DefaultUnreachableTolerationSeconds *int64 `json:"defaultUnreachableTolerationSeconds,omitempty" protobuf:"varint,15,opt,name=defaultUnreachableTolerationSeconds"`
+
+	//CORSAllowedOrigins contains a list of allowed origins for CORS, comma separated. An allowed origin can be a regular expression to support subdomain matching. If this list is empty CORS will not be enabled. (flag `--cors-allowed-origins`)
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_shoot.go]:
```suggestion
	// CORSAllowedOrigins contains a list of allowed origins for CORS, comma separated. An allowed origin can be a regular expression to support subdomain matching. If this list is empty CORS will not be enabled. (flag `--cors-allowed-origins`)
```

File: pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go [hunk 3]
```diff
@@ -165,6 +165,9 @@ type Values struct {
 	OIDC *gardencorev1beta1.OIDCConfig
 	// Requests contains configuration for the kube-apiserver requests.
 	Requests *gardencorev1beta1.KubeAPIServerRequests
+
```
Comment (rfranzke) [inline pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go]:
```suggestion
```

File: pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go [hunk 4]
```diff
@@ -165,6 +165,9 @@ type Values struct {
...
+	// CORSAllowedOrigins contans a list of List of allowed origins for CORS
```
Comment (rfranzke) [inline pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go]:
```suggestion
	// CORSAllowedOrigins contains a list of allowed origins for CORS.
```

File: pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go [hunk 5]
```diff
@@ -165,6 +165,9 @@ type Values struct {
...
+	CORSAllowedOrigins []string
```
Comment (rfranzke) [inline pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go:169]: Could you adapt the unit test to configure this field and ensure that the corresponding flag is set correctly?

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 6]
```diff
@@ -723,6 +723,10 @@ type KubeAPIServerConfig struct {
...
+	// +optional
+	CORSAllowedOrigins []string `json:"corsAllowedOrigins,omitempty" protobuf:"bytes,16,rep,name=corsAllowedOrigins"`
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_shoot.go:729]: Does it make sense to add some validation for this field, e.g., to ensure only valid domains are entered (maybe other restrictions, if needed)?

**Review by rfranzke — COMMENTED** [review]
Thanks for the PR!
/ok-to-test

> I would like to be able to consume the Shoot kubernetes API directly from a browser.

Can you perhaps explain a bit more about this? What does this mean/how does this consumption look like and why does it not work already today?

Generally, does it make sense to add some documentation about this into `docs/usage`?

**Comment by rfranzke** [comment]
Do you plan to continue with this PR?

**Comment by databus23** [comment]
Hi,

sorry I was busy with other stuff this week. Yes I intend to follow up but I also noticed that this PR only gets me half the way:
In order to access the kubernetes API from a browser I need the CORS headers but also a certificate that is deemed valid by the browser (e.g. signed by a public CA).
So for this to make any sense I would also need to be able to add an additional server certificate to the apiserver  (using `--tls-sni-cert-key`). It would also require reusing the existing wildcard letsencrypt certificate or provisioning a dedicated letsencrypt certificate for the apiserver.

This is more work then I anticipated. I will reassess this next week.

**Comment by oliver-goetz** [comment]
> Hi,
> 
> sorry I was busy with other stuff this week. Yes I intend to follow up but I also noticed that this PR only gets me half the way: In order to access the kubernetes API from a browser I need the CORS headers but also a certificate that is deemed valid by the browser (e.g. signed by a public CA). So for this to make any sense I would also need to be able to add an additional server certificate to the apiserver (using `--tls-sni-cert-key`). It would also require reusing the existing wildcard letsencrypt certificate or provisioning a dedicated letsencrypt certificate for the apiserver.
> 
> This is more work then I anticipated. I will reassess this next week.

PR #7712 already implemented using the existing wildcard certificate at kube-apiserver. It was just merged to master an hour ago. Thus, you can leverage that when you rebase.

**Comment by rfranzke** [comment]
cc @petersutter 
Do you have input/opinions for this PR?

**Comment by databus23** [comment]
> PR #7712 already implemented using the existing wildcard certificate at kube-apiserver. It was just merged to master an hour ago. Thus, you can leverage that when you rebase.

This is awesome! Thanks for the hint and what a timing coincidence.

**Comment by petersutter** [comment]
> cc @petersutter Do you have input/opinions for this PR?

not sure for what input you are looking for, however with the terminal feature in the gardener dashboard we do not have this issue as we are using a websocket connection from the browser to attach to the terminal pod, and CORS policy does not apply for websocket connections.

**Pull request closed without merging**